raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -592,17 +649,14 @@ private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
                 .setNextFetchOffset(nextOffsetAndEpoch.offset)
                 .setNextFetchOffsetEpoch(nextOffsetAndEpoch.epoch);
         } else if (quorum.isVoter(replicaId)) {
-            // Voters can read to the end of the log
             updateReplicaEndOffset(state, replicaId, fetchOffset);
-            state.highWatermark().ifPresent(log::updateHighWatermark);
-            Records records = log.read(fetchOffset, OptionalLong.empty());
-            return buildFetchQuorumRecordsResponse(Errors.NONE, records, highWatermark);
-        } else {
-            Records records = highWatermark.isPresent() ?
-                    log.read(fetchOffset, highWatermark) :
-                    MemoryRecords.EMPTY;
-            return buildFetchQuorumRecordsResponse(Errors.NONE, records, highWatermark);
         }
+
+        Records records = log.read(fetchOffset, OptionalLong.empty());
```

Comment:
Yes, that's right. Because we are propagating the high watermark through fetch responses, I could not think of a good reason observers should be limited to fetching below the high watermark